### PR TITLE
support Alibaba Cloud CDN storage middleware

### PR DESCRIPTION
**Pull request by Shawnpku**
Add support for Alibaba Cloud CDN (Content Delivery Network) when using OSS storage.

### CODE REVIEW

File: registry/storage/driver/middleware/alicdn/middleware.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	storagemiddleware "github.com/docker/distribution/registry/storage/driver/middleware"
+
+	"github.com/denverdino/aliyungo/cdn/auth"
+)
+
+// aliCdnStorageMiddleware provides a simple implementation of layerHandler that
+// constructs temporary signed AliCDN URLs from the storagedriver layer URL,
+// then issues HTTP Temporary Redirects to this AliCDN content URL.
+type aliCdnStorageMiddleware struct {
```
Comment (thaJeztah) [inline registry/storage/driver/middleware/alicdn/middleware.go]: Given that CDN is an acronym, I think this should be named `aliCDNStorageMiddleware` (see https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

File: registry/storage/driver/middleware/alicdn/middleware.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	duration  time.Duration
+}
+
+var _ storagedriver.StorageDriver = &aliCdnStorageMiddleware{}
+
+// newAliCdnLayerHandler constructs and returns a new AliCDN
+// LayerHandler implementation.
+// Required options: baseurl, authtype, privatekey
+// Optional options: duration
+func newAliCdnStorageMiddleware(storageDriver storagedriver.StorageDriver, options map[string]interface{}) (storagedriver.StorageDriver, error) {
```
Comment (thaJeztah) [inline registry/storage/driver/middleware/alicdn/middleware.go]: same here; `newAliCDNLayerHandler`

File: registry/storage/driver/middleware/alicdn/middleware.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+type aliCdnStorageMiddleware struct {
+	storagedriver.StorageDriver
+	baseURL   string
+	urlSigner *auth.URLSigner
+	duration  time.Duration
+}
+
+var _ storagedriver.StorageDriver = &aliCdnStorageMiddleware{}
+
+// newAliCdnLayerHandler constructs and returns a new AliCDN
```
Comment (thaJeztah) [inline registry/storage/driver/middleware/alicdn/middleware.go]: Looks like this doesn't match the actual function name

**Comment by Shawnpku** [comment]
@thaJeztah Thank you so much for the review. Fix them now.

**Comment by Shawnpku** [comment]
@thaJeztah Check the changes at your convenient. Thanks.

**Comment by thaJeztah** [comment]
Thanks for updating! I'm not a maintainer in this repository, so I'll leave the more in-depth review to the maintainers (I just spotted those small issues 🤗 )

**Comment by Shawnpku** [comment]
@dmcgowan request review. Thanks~

**Comment by thaJeztah** [comment]
oh! I see you missed a DCO sign-off on the second commit; could you 

- Squash your commits (no need to keep the second commit separate, as it was only a fix-up of the first one)
- Use your _real name_ for the sign-off (not your github handle)?

So

```
Signed-off-by: Shawn Chen <chen8132@gmail.com>
```

Instead of

```
Signed-off-by: Shawnpku <chen8132@gmail.com>
```

**Comment by Shawnpku** [comment]
Please take a look @dmcgowan @caervs @manishtomar. Thanks.

File: registry/storage/driver/middleware/alicdn/middleware.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	privateKey, ok := pk.(string)
+	if !ok {
+		return nil, fmt.Errorf("privatekey must be a string")
+	}
+
+	urlSigner := auth.NewURLSigner(authType, privateKey)
+
+	// parse duration
+	duration := 60 * time.Minute
```
Comment (caervs) [inline registry/storage/driver/middleware/alicdn/middleware.go]: For cloudfront we do 20 minute expiry. Any particular reason you think it should be longer for alibaba?

Comment (Shawnpku) [reply registry/storage/driver/middleware/alicdn/middleware.go]: @caervs I think 20 minute is ok. Do I need to change it?

**Comment by Shawnpku** [comment]
@caervs I think this is good now.

**Comment by Shawnpku** [comment]
@manishtomar would you like to take a review?

File: docs/configuration.md [hunk 5]
```diff
@@ -715,6 +715,17 @@ Then value of ipfilteredby:
...
 `awsregion`: IP from certain AWS regions goes to S3 directly, use together with `awsregion`
 
+### `alicdn`
+
+`alicdn` storage middleware allows the registry to serve layers via a content delivery network provided by Alibaba Cloud. Alicdn requires the OSS storage driver.
+
+| Parameter | Required | Description                                           |
+|-----------|----------|-------------------------------------------------------|
+| `baseurl` | yes      | The `SCHEME://HOST` at which Alicdn is served. |
+| `authtype` | yes   | The URL authentication type for Alicdn, which should be `a`, `b` or `c`.     |
```
Comment (manishtomar) [inline docs/configuration.md]: Could you please align the table? Also auth types seem weird. Can you provide CDN documentation link here that describes these auth types?

File: registry/storage/driver/middleware/alicdn/middleware.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	storagedriver.StorageDriver
+	baseURL   string
+	urlSigner *auth.URLSigner
+	duration  time.Duration
+}
+
+var _ storagedriver.StorageDriver = &aliCDNStorageMiddleware{}
+
+// newAliCDNStorageMiddleware constructs and returns a new AliCDN
+// layerHandler implementation.
```
Comment (manishtomar) [inline registry/storage/driver/middleware/alicdn/middleware.go]: I see you've taken the comment from cloudfront middleware but I think it should say `StorageDriver` implementation.

File: registry/storage/driver/middleware/alicdn/middleware.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
+package middleware
```
Comment (manishtomar) [inline registry/storage/driver/middleware/alicdn/middleware.go]: This should be `alicdn` as per golang convention.

**Comment by Shawnpku** [comment]
@manishtomar Thank you for the review. Changes applied.

**Comment by Shawnpku** [comment]
@dmcgowan @caervs @manishtomar Could someone merge this PR if it is good now. Thanks.

**Pull request merged**